### PR TITLE
No malloc for default alarm pool an pheap docs/cleanup

### LINKED ISSUE #129: Mixture of data types causing brain failure
When I created my Serial toolkit it took some time to get `int getchar_timeout_us (uint32_t timeout_us)` working.

I failed to spot the return type was int and as I was expecting a byte back, used uint8_t. Only after a bit of debugging (all praise the PicoProbe) did I figure out my mistake.

Additionally, the values for PICO_ERROR_GENERIC and PICO_ERROR_TIMEOUT are not documented in the SDK pdf so I ferreted them out of the source which was when I joined up the dots.

### PR DESCRIPTION
**Pull request by kilograham**

### CODE REVIEW

File: src/common/pico_time/time.c [hunk 1]
```diff
@@ -34,26 +34,25 @@ typedef struct alarm_pool {
 } alarm_pool_t;
 
 #if !PICO_TIME_DEFAULT_ALARM_POOL_DISABLED
-static alarm_pool_t *default_alarm_pool;
+// To avoid bringing in calloc, we statically allocate the arrays and he heap
```
Comment (lurch) [inline src/common/pico_time/time.c:37]: "he heap" -> "the heap" ?

File: src/common/pico_util/include/pico/util/pheap.h [hunk 2]
```diff
@@ -53,7 +53,11 @@ typedef struct pheap_node {
     pheap_node_id_t child, sibling, parent;
 } pheap_node_t;
 
-// return true if a < b in natural order
+/**
+ * A user comparator function for nodes in a pairing heap.
+ *
+ * \return return true if a < b in natural order. Note this relative ordering must be stable from call to call.
```
Comment (lurch) [inline src/common/pico_util/include/pico/util/pheap.h:59]: "\return return true" -> "\return true" ?

File: src/common/pico_util/include/pico/util/pheap.h [hunk 3]
```diff
@@ -67,17 +71,42 @@ typedef struct pheap {
     pheap_node_id_t free_tail_id;
 } pheap_t;
 
+/**
+ * Create a pairing heap, which effectively maintains an efficient sorted ordering
+ * of nodes. The heap itself stores no user per-node state, it is expected
+ * that th the user maintains a companion array. A comparator function must
```
Comment (lurch) [inline src/common/pico_util/include/pico/util/pheap.h:77]: "that th the" -> "that the"

File: src/common/pico_util/include/pico/util/pheap.h [hunk 4]
```diff
@@ -105,49 +135,154 @@ static pheap_node_id_t ph_merge_nodes(pheap_t *heap, pheap_node_id_t a, pheap_no
     }
 }
 
+/**
+ * Allocate a new node from the unused space in the heap
+ *
+ * \param heap the heap
+ * \return and identifier for the node, or 0 if the heap is full
```
Comment (lurch) [inline src/common/pico_util/include/pico/util/pheap.h:142]: "and identifier" -> "an identifier" ?

File: src/common/pico_util/include/pico/util/pheap.h [hunk 5]
```diff
@@ -105,49 +135,154 @@ static pheap_node_id_t ph_merge_nodes(pheap_t *heap, pheap_node_id_t a, pheap_no
...
     heap->root_id = ph_merge_nodes(heap, heap->root_id, id);
     return heap->root_id;
 }
 
+/**
+ * Returns the head node in the heap, i.e. the node
+ * which compares first, but without removing it from the heap.
+ *
+ * \param heap the heap
+ * \return the
```
Comment (lurch) [inline src/common/pico_util/include/pico/util/pheap.h:178]: return the ...?

File: src/common/pico_util/include/pico/util/pheap.h [hunk 6]
```diff
@@ -105,49 +135,154 @@ static pheap_node_id_t ph_merge_nodes(pheap_t *heap, pheap_node_id_t a, pheap_no
...
     return heap->root_id;
 }
 
-pheap_node_id_t ph_remove_head_reserve(pheap_t *heap, bool reserve);
+/**
+ * Remove the head node from the pairing heap. This head node is
+ * the node which compares first in the logical ordering provided
+ * by the comparator.
+ *
+ * Note that in the case of reserve = false, the returned id is no longer
```
Comment (lurch) [inline src/common/pico_util/include/pico/util/pheap.h:189]: "in the case of reserve = false" - does this need to be rewritten now that the parameter is named `free` ?

File: src/common/pico_util/include/pico/util/pheap.h [hunk 7]
```diff
@@ -105,49 +135,154 @@ static pheap_node_id_t ph_merge_nodes(pheap_t *heap, pheap_node_id_t a, pheap_no
...
+static inline pheap_node_id_t ph_remove_and_free_head(pheap_t *heap) {
+    return ph_remove_head(heap, true);
 }
 
-static inline bool ph_contains(pheap_t *heap, pheap_node_id_t id) {
+/**
+ * Remove and free an arbitrary node from the pairing heap. This is a more
+ * costly operation than removing the head via ph_remove_and_free_head()
+ *
+ * @param heap the heap
```
Comment (lurch) [inline src/common/pico_util/include/pico/util/pheap.h:221]: missing `@param id`

**Comment by kilograham** [comment]
fixes #126

**Review by liamfraser — APPROVED** [review]

**Review by liamfraser — APPROVED** [review]

**Pull request merged**